File: rust/src/core/discover.rs
use crate::config::EnvdoctorConfig;
use camino::Utf8PathBuf;
use crate::utils::glob::matches_glob;
use glob::glob;
use std::collections::HashSet;
use walkdir::WalkDir;

/// Always-ignored directories and files during discovery.
pub const ALWAYS_IGNORED: &[&str] = &[
    "node_modules",
    ".git",
    "dist",
    "build",
    ".next",
    ".turbo",
    ".vercel",
    ".netlify",
    "coverage",
    ".nyc_output",
    "target",
    ".cargo",
    "vendor",
    "Pods",
    ".idea",
    ".vscode",
    "*.log",
    "*.tmp",
    "*.swp",
    "*.swo",
    "~*",
    "*.DS_Store",
];

/// Git-aware file filter: tracks whether we're in a repo and can run `git
/// check-ignore` / `git diff` / `git ls-files`.
#[derive(Debug, Clone)]
pub struct GitFilter {
    pub root: Utf8PathBuf,
    pub is_repo: bool,
    pub tracked_files: HashSet<String>,
}

impl GitFilter {
    /// Create a new filter for the given root. Runs `git status` to check if
    /// it's a repo.
    pub fn new(root: &Utf8PathBuf) -> Self {
        let is_repo = Self::check_repo(root);
        let tracked_files = if is_repo {
            Self::list_tracked_files(root)
        } else {
            HashSet::new()
        };
        Self {
            root: root.clone(),
            is_repo,
            tracked_files,
        }
    }

    fn check_repo(root: &Utf8PathBuf) -> bool {
        std::process::Command::new("git")
            .args(["rev-parse", "--is-inside-work-tree"])
            .current_dir(root)
            .stdout(std::process::Stdio::null())
            .stderr(std::process::Stdio::null())
            .status()
            .map(|s| s.success())
            .unwrap_or(false)
    }

    fn list_tracked_files(root: &Utf8PathBuf) -> HashSet<String> {
        let output = std::process::Command::new("git")
            .args(["ls-files", "-z"])
            .current_dir(root)
            .output();
        match output {
            Ok(out) if out.status.success() => String::from_utf8_lossy(&out.stdout)
                .split('\0')
                .filter(|s| !s.is_empty())
                .map(|s| s.to_string())
                .collect(),
            _ => HashSet::new(),
        }
    }

    /// Should this file be skipped? Returns true to skip.
    pub fn should_skip(&self, path: &Utf8PathBuf) -> bool {
        // Check always-ignored patterns first (fast path)
        let file_name = path.file_name().unwrap_or("");
        for pattern in ALWAYS_IGNORED {
            if matches_glob(pattern, file_name) {
                return true;
            }
        }

        // If not a git repo, rely on patterns only
        if !self.is_repo {
            return false;
        }

        // Get relative path
        let rel = path.strip_prefix(&self.root).ok().and_then(|p| Some(p.as_str()));
        let Some(rel) = rel else { return false };

        // Skip if git ignores it (via .gitignore)
        let ignored = std::process::Command::new("git")
            .args(["check-ignore", "-q", rel])
            .current_dir(&self.root)
            .status()
            .map(|s| s.success())
            .unwrap_or(false);
        ignored
    }
}

/// Discover all files matching the configured glob patterns.
pub fn discover_files(root: &Utf8PathBuf, config: &EnvdoctorConfig) -> Vec<Utf8PathBuf> {
    let mut results = Vec::new();
    let patterns = [
        &config.env_file_patterns,
        &config.compose_file_patterns,
        &config.actions_file_patterns,
        &config.k8s_file_patterns,
    ];

    let git_filter = GitFilter::new(root);

    for pattern_group in &patterns {
        for pattern in *pattern_group {
            let full_pattern = root.join(pattern);
            if let Ok(paths) = glob(full_pattern.as_str()) {
                for entry in paths.flatten() {
                    let path = Utf8PathBuf::from_path_buf(entry).ok();
                    if let Some(p) = path {
                        if !git_filter.should_skip(&p) {
                            results.push(p);
                        }
                    }
                }
            }
        }
    }

    // Deduplicate and sort
    results.sort();
    results.dedup();
    results
}

/// Discover source files for usage scanning.
pub fn discover_source_files(root: &Utf8PathBuf, config: &EnvdoctorConfig) -> Vec<Utf8PathBuf> {
    let mut results = Vec::new();
    let git_filter = GitFilter::new(root);

    // Walk directory and match extensions
    for entry in WalkDir::new(root)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| e.file_type().is_file())
    {
        let path = Utf8PathBuf::from_path_buf(entry.path().to_path_buf()).ok();
        let Some(p) = path else { continue };

        if git_filter.should_skip(&p) {
            continue;
        }

        // Check extension
        let ext = p.extension().unwrap_or("");
        if config.source_extensions.iter().any(|e| e == ext) {
            results.push(p);
        }
    }

    results.sort();
    results.dedup();
    results
}

/// Get files changed since a commit/branch (for `diff` command).
#[allow(dead_code)]
pub fn changed_files_since(root: &Utf8PathBuf, since: &str) -> Vec<Utf8PathBuf> {
    let output = std::process::Command::new("git")
        .args(["diff", "--name-only", since])
        .current_dir(root)
        .output();
    match output {
        Ok(out) if out.status.success() => String::from_utf8_lossy(&out.stdout)
            .lines()
            .filter_map(|l| Some(root.join(l)))
            .collect(),
        _ => Vec::new(),
    }
}

/// Get currently staged files.
#[allow(dead_code)]
pub fn staged_files(root: &Utf8PathBuf) -> Vec<Utf8PathBuf> {
    let output = std::process::Command::new("git")
        .args(["diff", "--name-only", "--cached"])
        .current_dir(root)
        .output();
    match output {
        Ok(out) if out.status.success() => String::from_utf8_lossy(&out.stdout)
            .lines()
            .filter_map(|l| Some(root.join(l)))
            .collect(),
        _ => Vec::new(),
    }
}

File: dart/lib/src/core/discover.dart
/// File discovery for the audit pipeline.
library;

import 'dart:collection';
import 'dart:convert';
import 'dart:io';

import '../config/config.dart';
import '../utils/glob.dart';
import '../utils/paths.dart';

/// Always-ignored basenames during discovery.
const List<String> alwaysIgnored = [
  'node_modules',
  '.git',
  'dist',
  'build',
  '.next',
  '.turbo',
  '.vercel',
  '.netlify',
  'coverage',
  '.nyc_output',
  'target',
  '.cargo',
  'vendor',
  'Pods',
  '.idea',
  '.vscode',
  '*.log',
  '*.tmp',
  '*.swp',
  '*.swo',
  '~*',
  '*.DS_Store',
];

/// Directories that are never descended into.
final Set<String> ignoredDirs = {
  'node_modules',
  '.git',
  'dist',
  'build',
  'coverage',
  '.next',
  '.nuxt',
  '.venv',
  'vendor',
  '.Trash',
  'Library',
  '.cache',
  '.npm',
  '.turbo',
  '.vercel',
  '.netlify',
  '.nyc_output',
  'target',
  '.cargo',
  'Pods',
  '.idea',
  '.vscode',
};

/// Git-aware file filter: tracks whether we're in a repo and can run
/// `git check-ignore` / `git diff`.
class GitFilter {
  final String root;
  final bool isRepo;

  GitFilter(this.root) : isRepo = _checkRepo(root);

  static bool _checkRepo(String root) {
    try {
      final result = Process.runSync('git', ['rev-parse', '--is-inside-work-tree'],
          workingDirectory: root, stdoutEncoding: utf8, stderrEncoding: utf8);
      return result.exitCode == 0;
    } catch (_) {
      return false;
    }
  }

  /// Should this file be skipped? Returns true to skip.
  bool shouldSkip(String path) {
    // Check always-ignored patterns first (fast path).
    final fileName = basename(path);
    for (final pattern in alwaysIgnored) {
      if (matchesGlob(pattern, fileName)) return true;
    }

    // If not a git repo, rely on patterns only.
    if (!isRepo) return false;

    final rel = relativeTo(root, path);
    if (rel == null) return false;

    // Skip if git ignores it (via .gitignore).
    try {
      final result = Process.runSync('git', ['check-ignore', '-q', rel],
          workingDirectory: root, stdoutEncoding: utf8, stderrEncoding: utf8);
      return result.exitCode == 0;
    } catch (_) {
      return false;
    }
  }
}

/// Discover all files matching the configured glob patterns.
List<String> discoverFiles(String root, EnvdoctorConfig config) {
  final patterns = [
    ...config.envFilePatterns,
    ...config.composeFilePatterns,
    ...config.actionsFilePatterns,
    ...config.k8sFilePatterns,
  ].map(pathGlobToRegExp).toList();

  final gitFilter = GitFilter(root);
  final results = <String>[];

  for (final path in walkFiles(root)) {
    final rel = relativeTo(root, path);
    if (rel == null) continue;
    if (!patterns.any((re) => re.hasMatch(rel))) continue;
    if (gitFilter.shouldSkip(path)) continue;
    results.add(path);
  }

  results.sort();
  return LinkedHashSet<String>.from(results).toList();
}

/// Discover source files for usage scanning.
List<String> discoverSourceFiles(String root, EnvdoctorConfig config) {
  final extensions = config.sourceExtensions
      .map((e) => e.startsWith('.') ? e.substring(1) : e)
      .toSet();
  final gitFilter = GitFilter(root);
  final results = <String>[];

  for (final path in walkFiles(root)) {
    if (gitFilter.shouldSkip(path)) continue;
    final ext = extensionOf(path);
    if (ext.isNotEmpty && extensions.contains(ext)) results.add(path);
  }

  results.sort();
  return LinkedHashSet<String>.from(results).toList();
}

/// Walk every file under root, pruning always-ignored directories.
Iterable<String> walkFiles(String root) sync* {
  final pending = <String>[root];
  while (pending.isNotEmpty) {
    final dir = pending.removeLast();
    List<FileSystemEntity> entries;
    try {
      entries = Directory(dir).listSync(followLinks: false).toList();
    } catch (_) {
      continue;
    }
    for (final entry in entries) {
      final name = basename(entry.path);
      if (entry is Directory) {
        if (!ignoredDirs.contains(name)) pending.add(entry.path);
      } else if (entry is File) {
        yield entry.path;
      }
    }
  }
}

/// Get files changed since a commit/branch (including untracked files).
List<String> changedFilesSince(String root, String since) =>
    _gitNameOnly(root, ['diff', '--name-only', since], includeUntracked: true);

/// Get currently staged files.
List<String> stagedFiles(String root) =>
    _gitNameOnly(root, ['diff', '--name-only', '--cached'], includeUntracked: false);

List<String> _gitNameOnly(String root, List<String> args,
    {required bool includeUntracked}) {
  try {
    final topLevelResult = Process.runSync(
      'git',
      ['rev-parse', '--show-toplevel'],
      workingDirectory: root,
      stdoutEncoding: utf8,
      stderrEncoding: utf8,
    );
    if (topLevelResult.exitCode != 0) return [];
    final topLevel = topLevelResult.stdout.trim();
    final result = Process.runSync('git', ['-C', topLevel, ...args],
        stdoutEncoding: utf8, stderrEncoding: utf8);
    if (result.exitCode != 0) return [];
    final files = <String>[];
    final seen = <String>{};
    void addResolved(String line) {
      final trimmed = line.trim();
      if (trimmed.isEmpty) return;
      final full = normalizePath(
          isAbsolutePath(trimmed) ? trimmed : joinPath(topLevel, trimmed));
      if (seen.add(full)) files.add(full);
    }

    for (final line in (result.stdout as String).split('\n')) {
      addResolved(line);
    }
    if (includeUntracked) {
      try {
        final untracked = Process.runSync(
          'git',
          ['-C', topLevel, 'ls-files', '--others', '--exclude-standard'],
          workingDirectory: root,
          stdoutEncoding: utf8,
          stderrEncoding: utf8,
        );
        if (untracked.exitCode == 0) {
          for (final line in (untracked.stdout as String).split('\n')) {
            addResolved(line);
          }
        }
      } catch (_) {
        // ignore
      }
    }
    return files;
  } catch (_) {
    return [];
  }
}

/// True when git-aware filtering is active but no git changes were found.
bool hasNoGitChanges(String root, {bool? staged, String? since}) {
  if (staged == true) {
    return stagedFiles(root).isEmpty;
  }
  if (since != null) {
    return changedFilesSince(root, since).isEmpty;
  }
  return false;
}

final Map<String, RegExp> _pathGlobCache = {};

/// Translate a path glob (with `**`, `*`, `?`, and `{a,b}` braces) to a regex
/// matched against a root-relative, forward-slash path.
RegExp pathGlobToRegExp(String pattern) {
  final cached = _pathGlobCache[pattern];
  if (cached != null) return cached;
  final sb = StringBuffer('^');
  var i = 0;
  while (i < pattern.length) {
    final c = pattern[i];
    switch (c) {
      case '*':
        if (i + 1 < pattern.length && pattern[i + 1] == '*') {
          // `**/`: zero or more whole segments; bare `**`: anything.
          if (i + 2 < pattern.length && pattern[i + 2] == '/') {
            sb.write('(?:.*/)?');
            i += 3;
          } else {
            sb.write('.*');
            i += 2;
          }
        } else {
          sb.write('[^/]*');
          i += 1;
        }
      case '?':
        sb.write('[^/]');
        i += 1;
      case '{':
        final close = pattern.indexOf('}', i + 1);
        if (close < 0) {
          sb.write(r'\{');
          i += 1;
        } else {
          final inner = pattern.substring(i + 1, close);
          final options = inner.split(',').map(RegExp.escape);
          sb.write('(?:${options.join('|')})');
          i = close + 1;
        }
      default:
        sb.write(RegExp.escape(c));
        i += 1;
    }
  }
  sb.write(r'$');
  final re = RegExp(sb.toString());
  _pathGlobCache[pattern] = re;
  return re;
}
